src/common/tools/merge_listing_times.ts

```diff
@@ -0,0 +1,177 @@
+only increase the values in the listing_meta.json file, and will only cause WPT
+chunks to become smaller.
+
+If there are no TIMING_LOG_FILES, this just regenerates (reformats) the file
+using the data already present.
+
+In more detail:
+
+- Reads per-case timing data in any of the SUITES, from all TIMING_LOG_FILES
+  (ignoring skipped cases), and averages it over the number of subcases.
+  In the case of cases that have run multiple times, takes the max of each.
+- Compiles the average time-per-subcase for each test seen.
+- For each suite seen, loads its listing_meta.json, takes the max of the old and
+  new data, and writes it back out.
+
+How to generate TIMING_LOG_FILES files:
+
+- Launch the 'websocket-logger' tool (see its README.md), which listens for
+  log messages on localhost:59497.
+- Run the tests you want to capture data for, on the same system. Since
+  logging is done through the websocket side-channel, you can run the tests
+  under any runtime (standalone, WPT, etc.) as long as WebSocket support is
+  available (always true in browsers).
+- Run \`tools/merge_listing_times webgpu -- tools/websocket-logger/wslog-*.txt\`
+`);
+  process.exit(rc);
+}
+
+const kHeader = `{
+  "_comment": "SEMI AUTO-GENERATED: Please read tools/merge_listing_times.",
+`;
+const kFooter = `\
+  "_end": ""
+}
+`;
+
+const argv = process.argv;
+if (argv.some(v => v.startsWith('-') && v !== '--') || argv.every(v => v !== '--')) {
+  usage(0);
+}
+const suites = [];
+const timingLogFilenames = [];
+let seenDashDash = false;
+for (const arg of argv.slice(2)) {
+  if (arg === '--') {
+    seenDashDash = true;
+    continue;
+  } else if (arg.startsWith('-')) {
+    usage(0);
+  }
+
+  if (seenDashDash) {
+    timingLogFilenames.push(arg);
+  } else {
+    suites.push(arg);
+  }
+}
+if (!seenDashDash) {
+  usage(0);
+}
+
+void (async () => {
+  // Read the log files and find the entry for each `q`
+  // that has the largest average subcase time.
+  const caseTimes = new Map<string, CaseTimingLogLine>();
+  for (const timingLogFilename of timingLogFilenames) {
+    const rl = readline.createInterface({
+      input: fs.createReadStream(timingLogFilename),
+      crlfDelay: Infinity,
+    });
+
+    for await (const line of rl) {
+      const parsed: CaseTimingLogLine = JSON.parse(line);
+
+      const prev = caseTimes.get(parsed.q);
+      if (prev !== undefined) {
+        const timePerSubcase = parsed.timems / Math.max(1, parsed.unskippedSubcaseCount);
+        const prevTimePerSubcase = prev.timems / Math.max(1, prev.unskippedSubcaseCount);
+
+        if (timePerSubcase > prevTimePerSubcase) {
+          caseTimes.set(parsed.q, parsed);
+        }
+      } else {
+        caseTimes.set(parsed.q, parsed);
+      }
+    }
+  }
+
+  // Accumulate total times per test. Map of suite -> query -> {totalTimeMS, caseCount}.
+  const testTimes = new Map<string, Map<string, { totalTimeMS: number; subcaseCount: number }>>();
```

Comment:
doesn't the query have the suite in it? why do we need suite -> query ?

you can still check below `if (!suites.includes(castQ.suite)) continue;`

Reply:
It was just to optimize that but that's a silly optimization because we basically never use more than one suite. done

Reply:
Hmm actually it's not quite so simple. The loop after that, which updates the metadata, has to sort the results into which metadata file they go in, which requires knowing the suite from just a query string. It's slightly annoying to parse the suite out so I'm just going to leave this as is...